commit 7c2061bdfba7c738dac1e2c14db51faeef3f34b1
Author: Heikki Linnakangas <heikki.linnakangas@iki.fi>
Date:   Fri Dec 5 23:39:01 2025 +0200

    Fix test to work with non-8kB block sizes
    
    Author: Maxim Orlov <orlovmg@gmail.com>
    Discussion: https://www.postgresql.org/message-id/CACG%3Dezbtm%2BLOzEMyLX7rzGcAv3ez3F6nNpSJjvZeMzed0Oe6Pw%40mail.gmail.com

diff --git a/src/test/modules/test_slru/t/002_multixact_wraparound.pl b/src/test/modules/test_slru/t/002_multixact_wraparound.pl
index de37d845b11..169333fc564 100644
--- a/src/test/modules/test_slru/t/002_multixact_wraparound.pl
+++ b/src/test/modules/test_slru/t/002_multixact_wraparound.pl
@@ -23,23 +23,35 @@ command_ok(
 		'--multixact-ids' => '0xFFFFFFF8,0xFFFFFFF8',
 		$node_pgdata
 	],
 	"set the cluster's next multitransaction to 0xFFFFFFF8");
 
+# Extract a few values from pg_resetwal --dry-run output that we need for
+# the calculations below
+my $out = (run_command([ 'pg_resetwal', '--dry-run', $node->data_dir ]))[0];
+$out =~ /^Database block size: *(\d+)$/m or die;
+my $blcksz = $1;
+$out =~ /^Pages per SLRU segment: *(\d+)$/m or die;
+my $slru_pages_per_segment = $1;
+
 # Fixup the SLRU files to match the state we reset to.
 
-# initialize SLRU file with zeros (65536 entries * 4 bytes = 262144 bytes)
-my $slru_file = "$node_pgdata/pg_multixact/offsets/FFFF";
+# initialize the 'offsets' SLRU file containing the new next multixid
+# with zeros
+my $multixact_offsets_per_page = $blcksz / 4;   # sizeof(MultiXactOffset) == 4
+my $segno =
+  int(0xFFFFFFF8 / $multixact_offsets_per_page / $slru_pages_per_segment);
+my $slru_file = sprintf('%s/pg_multixact/offsets/%04X', $node_pgdata, $segno);
 open my $fh, ">", $slru_file
   or die "could not open \"$slru_file\": $!";
 binmode $fh;
-# Write 65536 entries of 4 bytes each (all zeros)
-syswrite($fh, "\0" x 262144) == 262144
+my $bytes_per_seg = $slru_pages_per_segment * $blcksz;
+syswrite($fh, "\0" x $bytes_per_seg) == $bytes_per_seg
   or die "could not write to \"$slru_file\": $!";
 close $fh;
 
-# remove old SLRU file
+# remove old file
 unlink("$node_pgdata/pg_multixact/offsets/0000")
   or die "could not unlink \"$node_pgdata/pg_multixact/offsets/0000\": $!";
 
 # Consume multixids to wrap around.  We start at 0xFFFFFFF8, so after
 # creating 16 multixacts we should definitely have wrapped around.
